### PR TITLE
policy/k8s: Fix bug where policy synchronization event was lost

### LINKED ISSUE #31865: bug: endpoint regeneration proceeds before policy synchronization has completed
I was working on another change when I noticed this.

Likely as a result of #30322, I noticed endpoints undergoing their first regeneration after agent restart before policy had been fully applied. This is unfortunate, as it meant endpoints briefly had incorrect policy.

I added some additional logging and saw something like this:

```
2024-04-09T10:35:53.379194584Z time="2024-04-09T10:35:53Z" level=debug msg="resource \"cilium/v2::CiliumNetworkPolicy\" cache has synced, stopping timeout watcher" subsys=k8s
2024-04-09T10:36:01.638609794Z time="2024-04-09T10:36:01Z" level=info msg="CacheStatus closed, proceeding with regeneration" subsys=daemon
2024-04-09T10:36:01.687707197Z time="2024-04-09T10:36:01Z" level=debug msg="waiting for cache to synchronize" kubernetesResource="cilium/v2::CiliumNetworkPolicy" subsys=k8s
2024-04-09T10:36:01.687726580Z time="2024-04-09T10:36:01Z" level=debug msg="Is CNP synced?" subsys=policy-k8s-watcher v=false
2024-04-09T10:36:01.691903550Z time="2024-04-09T10:36:01Z" level=debug msg="Adding CiliumNetworkPolicy" ciliumNetworkPolicyName=allow-one-fqdn-http k8sApiVersion=cilium.io/v2 k8sNamespace=foo subsys=policy-k8s-watcher
2024-04-09T10:36:01.692351378Z time="2024-04-09T10:36:01Z" level=info msg="Policy Add Request" ciliumNetworkPolicy= [snip]
2024-04-09T10:36:01.692450366Z time="2024-04-09T10:36:01Z" level=info msg="Imported CiliumNetworkPolicy" ciliumNetworkPolicyName=allow-one-fqdn-http k8sApiVersion=cilium.io/v2 k8sNamespace=foo subsys=policy-k8s-watcher
2024-04-09T10:36:01.788546485Z time="2024-04-09T10:36:01Z" level=debug msg="Is CNP synced?" subsys=policy-k8s-watcher v=false
2024-04-09T10:36:01.788585344Z time="2024-04-09T10:36:01Z" level=debug msg="CNP sync complete" subsys=policy-k8s-watcher
```

I also saw the endpoint in question undergo two regenerations, one of which definitely was before the PolicyRepository was fully populated.

### PR DESCRIPTION
**Pull request by gandro**
This PR fixes an issue with the K8s policy ingestor where `InitK8sSubsystem` unblocked the daemon startup before all initial policies were processed by the ingestor.

This meant that we created unnecessary endpoint regenerations at startup, due to the initial endpoint regeneration being disrupted by the discovery of additional policies (and thus re-triggering endpoint regeneration).

The cause for the race was that we started the K8s policy watcher in the Hive lifecycle start hook, which is too late: The `InitK8sSubsystem` function called from newDaemon constructor is supposed to block on policy resources being synced. But because we registered those resources (via `BlockWaitGroupToSyncResources`) only after the `newDaemon` constructor already ran (we depended on a fully constructed `Daemon` object), `InitK8sSubsystem` would continue without the policy resources being synced and thus causing unnecessary endpoint regenerations.

This PR fixes the issue by starting the policy resource watcher earlier, namely directly in `newDaemon`. This unfortunately means that we have to side-step the Hive infrastructure, as we currently require a partially initialized `Daemon` object to implement the `PolicyManager` interface for the time being. Hopefully, in the future we can move the `PolicyManager` logic out of the `Daemon` struct and thereby breaking this cyclic dependency.

I have manually tested this, both by checking the logs and by disabling the CNP sync manually (which then blocks the agent before it restores endpoints). Unfortunately, I didn't find a good way to assert this via unit tests.

Fixes: https://github.com/cilium/cilium/issues/31865

### CODE REVIEW

File: pkg/policy/k8s/cell.go [hunk 1]
```diff
@@ -66,63 +65,63 @@ type PolicyWatcherParams struct {
...
+	// (i.e. watchResources has observed the Sync event for CNP/CCNP/KNP).
+	// Because the PolicyManager interface itself is implemented by the Daemon
+	// struct, we have a circular dependency.
+
+	// In order to not miss any service events, we register a subscriber here,
+	// before the call to K8sWatcher.InitK8sSubsystem.
+	svcCacheNotifications := stream.ToChannel(ctx, p.params.ServiceCache.Notifications(),
+		stream.WithBufferSize(int(p.params.Config.K8sServiceCacheSize)))
+
+	w := &policyWatcher{
```
Comment (derailed) [inline pkg/policy/k8s/cell.go:112]: nit: Perhaps we could use a ctor here and hide the semantic + opts at the call site?

**Review by derailed — COMMENTED** [review]
@gandro Nice work!

File: pkg/policy/k8s/cell.go (hunk 1 above)
Comment (gandro) [reply pkg/policy/k8s/cell.go:112]: Thanks. I've moved the code to a separate constructor. Is this what you had in mind?

**Comment by squeed** [comment]
Tagged sig/foundations here.

**Review by squeed — APPROVED** [review]

**Review by dylandreimerink — APPROVED** [review]
Like discussed out of band, this isn't a great permanent solution, but will approve for now and work on proper long term fix later

**Pull request merged**